README.md
# Deploy Beats
[![Build Status](https://travis-ci.com/weesntos/deploybeats.svg?branch=master)](https://travis-ci.com/weesntos/deploybeats)

# Como usar

## Pre-requisito

* Instalar o **Git** para download do `codigo fonte`. [Clique aqui](https://git-scm.com/downloads)
* Instalar o **Docker** para hospedar os `containers`. [Windows](https://docs.docker.com/docker-for-windows/install/) | [MacOS](https://docs.docker.com/docker-for-mac/install/) | [Linux](https://docs.docker.com/install/linux/docker-ce/ubuntu/)
* Instalar o **Docker-Compose** para `build` e `init` das aplicacoes. [Clique aqui](https://docs.docker.com/compose/install/)


![stack](stack.png)

## 1 - Download do codigo fonte

```
git clone https://github.com/weesntos/deploybeats.git
```

## 2 - Build

### Stack - Frontend + Backend
```
docker-compose build
```

### Frontend Only
```
docker-compose build frontend
```

### Backend Only
```
docker-compose build backend
```

## 3 - Init

### Stack - Frontend + Backend
```
docker-compose up -d
docker-compose logs -f
```

### Frontend Only
```
docker-compose up -d --no-deps frontend
docker-compose logs -f
```

### Backend Only
```
docker-compose up -d backend
docker-compose logs -f
```

## 4 - Abrir e Testar Aplicacao
WEB: http://localhost:8000

API: http://localhost:8001

```
docker-compose run tests
```


## 5 - Destroy
```
docker-compose down
```

## All In One: Build + Destroy + Init
```
docker-compose up --build --force-recreate -d
docker-compose logs -f
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.1] - 2019-10-16
### Changed
- update changelog

## [2.0.0] - 2019-10-16
### Changed
- support python 3.7.4

## [1.0.0] - 2019-10-09
### Added
- Deploy App Heroku
- Version navigation.
- Source code
- Docker-compose


src/backend/backend.py
#!/usr/bin/env python3
import datetime
import os
import pymongo
from bson import ObjectId
from flask import make_response
from flask import Flask, jsonify
from flask_restful import reqparse
from pymongo import MongoClient
from bson.json_util import dumps

app = Flask(__name__)
app_port = os.getenv("PORT", 8001)

# Mongo DB Connection
db_uri = os.getenv("DB_URI", "mongodb://localhost")
client = MongoClient(db_uri)
db = client.mydb
collection = db.deploys

# Parser
parser = reqparse.RequestParser()
parser.add_argument('componente')
parser.add_argument('versao')
parser.add_argument('responsavel')
parser.add_argument('status')


@app.route('/api/deploys', methods=['GET'])
def get():
    try:
        query = collection.find().sort('_id', pymongo.DESCENDING)
        docs = dumps(query)
        app.logger.info("{}".format(docs))
        return (docs)
    except Exception as e:
        app.logger.error(e)


@app.route('/api/deploys/<id>', methods=['GET'])
def get_deploy(id):
    try:
        query = collection.find({"_id": ObjectId(id)})
        docs = dumps(query)
        app.logger.info("{}".format(docs))
        return (docs)
    except Exception as e:
        app.logger.error(e)


@app.route('/api/deploys', methods=['POST'])
def post():
    try:
        args = parser.parse_args()
        date = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        doc = {'componente': args['componente'],
               'versao': args['versao'],
               'responsavel': args['responsavel'],
               'status': args['status'],
               'criado': f"{date}",
               'atualizado': ''}
        res = collection.insert_one(doc).inserted_id
        app.logger.info("ID: {}, Payload: {}".format(res, doc))
        return jsonify({"status_code": "201", "id": "{}".format(res)}), 201
    except Exception as e:
        app.logger.error(e)


@app.route('/api/deploys/<id>', methods=['PUT'])
def put(id):
    try:
        args = parser.parse_args()
        date = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        collection.update_one({"_id": ObjectId(id)},
                              {"$set": {'status': args['status'],
                                        'atualizado': f"{date}"}})
        app.logger.info("ID: {}".format(id))
        return jsonify({"status_code": "200", "id": "{}".format(id)})
    except Exception as e:
        app.logger.error(e)


@app.errorhandler(404)
def not_found(error):
    return make_response(jsonify({'error': 'Not found'}), 404)


if __name__ == '__main__':
    app.run(debug=True, host="0.0.0.0", port=app_port)


src/frontend/frontend.py
import os
import requests
from flask import Flask, request, jsonify, render_template

app = Flask(__name__)
app_port = os.getenv("PORT", 8000)
backend_host = os.getenv("BACKEND_HOST", "localhost")
backend_port = os.getenv("BACKEND_PORT", 8001)


@app.route('/', methods=['GET', 'POST'])
def index():
    template = "index.html"
    app.logger.info(template)
    if request.method == 'POST':
        componente = request.form.get("componente")
        responsavel = request.form.get("responsavel")
        status = request.form.get("status")
        versao = request.form.get("versao")
        try:
            status_code, response = create_deploy(
                componente, responsavel, status, versao)
            return render_template(template,
                                   status_code=status_code,
                                   response=response)
        except Exception as e:
            app.logger.error(e)

    return render_template(template)


@app.route('/deploys', methods=['GET'])
def view_deploys():
    template = "view_deploys.html"
    try:
        status_code, response = list_deploys()
        return render_template(template,
                               status_code=status_code,
                               response=response)
    except Exception as e:
        app.logger.error(e)


@app.route('/health', methods=['GET'])
def health():
    return jsonify({"frontend_status": "ok",
                    "backend_status": "{}".format(backend_status())})


def backend_status():
    try:
        url = "http://{}:{}/health".format(backend_host, backend_port)
        app.logger.info(url)
        return "ok"
    except Exception as e:
        app.logger.error(e)
        return "nok"


def create_deploy(componente, responsavel, status, versao):
    app.logger.info("Recebido: componente: %s, versao: %s", componente, versao)
    try:
        url = "http://{}:{}/api/deploys".format(backend_host, backend_port)
        json = {"componente": f"{componente}",
                "responsavel": f"{responsavel}",
                "status": f"{status}",
                "versao": f"{versao}"}
        app.logger.info("Url {}, Payload {}".format(url, json))

        r = requests.post(url, json=json)
        if r.status_code == 201:
            status_code, response = r.status_code, r.json()
        else:
            status_code, response = r.status_code, None

        app.logger.info("Result code {}, Reponse {}".format(
            status_code, response))
        return status_code, response

    except Exception as e:
        app.logger.error(e)


def list_deploys():
    try:
        url = "http://{}:{}/api/deploys".format(backend_host, backend_port)
        r = requests.get(url)
        if r.status_code == 200:
            status_code, response = r.status_code, r.json()
        else:
            status_code, response = r.status_code, None
        return status_code, response
    except Exception as e:
        app.logger.error(e)


if __name__ == "__main__":
    app.run(debug=True, host="0.0.0.0", port=app_port)
